File: recipes/subms-health/python/src/subms_health/clock.py
"""Clock seam: wall-clock millis (for staleness) + an RFC3339 stamp (for the
report's refreshed_at). Tests freeze it to pin cross-language fixtures."""

from __future__ import annotations

import time


class Clock:
    def now_ms(self) -> int:  # pragma: no cover - abstract
        raise NotImplementedError

    def now_rfc3339(self) -> str:  # pragma: no cover - abstract
        raise NotImplementedError


class SystemClock(Clock):
    def now_ms(self) -> int:
        return int(time.time() * 1000)

    def now_rfc3339(self) -> str:
        return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


class FixedClock(Clock):
    def __init__(self, ms: int, rfc3339: str) -> None:
        self._ms = ms
        self._stamp = rfc3339

    def set(self, ms: int) -> None:
        self._ms = ms

    def now_ms(self) -> int:
        return self._ms

    def now_rfc3339(self) -> str:
        return self._stamp

File: recipes/subms-health/java/src/main/java/com/submillisecond/recipes/health/Clock.java
package com.submillisecond.recipes.health;

import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;

/** Clock seam: wall-clock millis (staleness) + RFC3339 stamp (refreshed_at). */
public interface Clock {
    long nowMs();

    String nowRfc3339();

    final class SystemClock implements Clock {
        private static final DateTimeFormatter FMT =
                DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss'Z'").withZone(ZoneOffset.UTC);

        @Override
        public long nowMs() {
            return System.currentTimeMillis();
        }

        @Override
        public String nowRfc3339() {
            return FMT.format(Instant.now());
        }
    }

    final class FixedClock implements Clock {
        private long ms;
        private final String stamp;

        public FixedClock(long ms, String rfc3339) {
            this.ms = ms;
            this.stamp = rfc3339;
        }

        public void set(long ms) {
            this.ms = ms;
        }

        @Override
        public long nowMs() {
            return ms;
        }

        @Override
        public String nowRfc3339() {
            return stamp;
        }
    }
}
